src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,78 +34,236 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
+     * not already in {@code bestMatchingMilestones}
      *
-     * @return
+     * This method will do nothing if {@code limit} is not a positive integer
+     * @param bestMatchingMilestones
+     * @param limit
+     * @param userInput
+     * @param milestones
      */
-    public List<PickerMilestone> getMatchingMilestonesList() {
-        return this.currentMilestonesList.stream()
-                .filter(milestone -> !milestone.isMatching())
+    private void addPreviouslyMatchingMilestones(List<PickerMilestone> bestMatchingMilestones, int limit,
+                                                 String userInput, List<PickerMilestone> milestones) {
+        if (limit <= 0) return;
+        List<PickerMilestone> previouslyMatchingMilestones =
+                getLimitedMatchesFromPreviousInput(userInput, bestMatchingMilestones, limit, milestones);
+        bestMatchingMilestones.addAll(previouslyMatchingMilestones);
+    }
+
+    /**
+     * Adds up to {@code limit} number of milestones to {@code bestMatchingMilestones}
+     * These milestones are taken from {@code milestones} if its isMatching() returns true
+     *
+     * This method will do nothing if {@code limit} is not a positive integer
+     * @param bestMatchingMilestones
+     * @param limit
+     * @param milestones
+     */
+    private void addMatchingMilestones(List<PickerMilestone> bestMatchingMilestones, int limit,
+                                       List<PickerMilestone> milestones) {
+        if (limit <= 0) return;
+        List<PickerMilestone> matchingMilestones = getLimitedMatchingMilestones(limit, milestones);
+        bestMatchingMilestones.addAll(matchingMilestones);
+    }
+
+    /**
+     * Gets first {@code limit} milestones from {@code milestones} which are not already in {@code allMilestones}
+     *
+     * Assumption: {@code limit} is not negative
+     * @param currentMilestones
+     * @param limit
+     * @param milestones
+     */
+    private List<PickerMilestone> getLimitedFirstMilestones(List<PickerMilestone> currentMilestones,
+                                                            int limit, List<PickerMilestone> milestones) {
+        return milestones.stream()
+                .filter(milestone -> !currentMilestones.contains(milestone))
+                .limit(limit)
                 .collect(Collectors.toList());
     }
 
-    private void filterMilestones(String query) {
-        currentMilestonesList
-                .forEach(milestone -> {
-                    boolean matchQuery = Utility.containsIgnoreCase(milestone.getTitle(), query);
-                    if (!milestone.isMatching()) milestone.setMatching(!matchQuery);
-                });
+    /**
+     * Concatenates all the matching milestones for previous userInput(s), then gets the first {@code limit}
+     * milestones which are not already in {@code currentMilestones}
```

Comment:
I think the comment can be better. The implementation of doing concatenation is not important to be mentioned in the comment. Maybe `Gets a limited number of milestones matching previous userInput(s) which are not already in {@code currentMilestones}.`